src/support/generation.c

```diff
@@ -159,14 +162,31 @@ __wt_gen_drain(WT_SESSION_IMPL *session, int which, uint64_t generation)
              */
             if (minutes == 0) {
                 minutes = 1;
-                __wt_seconds(session, &start);
+                __wt_epoch(session, &start);
             } else {
-                __wt_seconds(session, &stop);
-                if (stop - start > minutes * WT_MINUTE) {
+                __wt_epoch(session, &stop);
+                time_diff = WT_TIMEDIFF_MS(stop, start);
+#define WT_GEN_DRAIN_TIMEOUT_MIN 4
+                if (time_diff > minutes * WT_MINUTE * WT_THOUSAND) {
                     __wt_verbose_notice(session, WT_VERB_GENERATION,
                       "%s generation drain waited %u minutes", __gen_name(which), minutes);
                     ++minutes;
-                    WT_ASSERT(session, minutes < 4);
+                    WT_ASSERT(session, minutes < WT_GEN_DRAIN_TIMEOUT_MIN);
+                }
+                /* Enable extra logs 20ms before timing out. */
+                else if (!verbose_timeout_flags &&
+                  time_diff > ((WT_GEN_DRAIN_TIMEOUT_MIN - 1) * WT_MINUTE * WT_THOUSAND - 20)) {
```

Comment:
should this be:
`time_diff > (WT_GEN_DRAIN_TIMEOUT_MIN * WT_MINUTE * WT_THOUSAND - 20)`
i.e.
`time_diff > (4 * 60 * 1000 - 20)`

Reply:
Thanks, fixed in a86cb05b9fb142cff2d8fb07a51791b6a1b30f60